components/input/Input.js

```diff
@@ -73,42 +83,91 @@ export default class Input extends Component {
   }
 
   handleIncreaseValue() {
-    this.setState((previousState, props) => ({
-      value: previousState.value ? previousState.value + props.step : props.step,
-    }));
+    this.step(1);
   }
 
   handleDecreaseValue() {
-    this.setState((previousState, props) => ({
-      value: previousState.value ? previousState.value - props.step : -props.step,
-    }));
+    this.step(-1);
+  }
+
+  hasError() {
+    const { meta } = this.props;
+    return Boolean(meta && meta.error && meta.touched);
+  }
+
+  format(number) {
+    const { precision } = this.props;
+
+    let formattedNumber = this.toNumber(number);
+
+    if (precision !== null) {
+      formattedNumber = number.toFixed(precision);
+    }
+
+    formattedNumber += '';
+
+    return formattedNumber;
+  }
+
+  step(n) {
+    const { step } = this.props;
+
+    const number = this.toNumber((this.state.value || 0) + step * n);
+
+    if (number !== this.state.value) {
+      this.setState({ value: number });
+      return true;
+    }
+
+    return false;
+  }
+
+  toNumber(number) {
+    const { max, min, precision } = this.props;
+
+    let float = parseFloat(number);
+
+    if (isNaN(float) || !isFinite(float)) {
+      float = 0;
+    }
+
+    const q = Math.pow(10, precision === null ? 10 : precision);
+
+    float = Math.min(Math.max(float, min), max);
+    float = Math.round(float * q) / q;
+
+    return float;
```

Comment:
I feel like most of these helper functions don't really belong in (this) component.